test/include/seqan3/test/tmp_directory.hpp

```diff
@@ -0,0 +1,210 @@
+        // returned by temp_directory_path. Within this path we can safely create files, that would be
+        // unique per test instance as the parent directory is.
+        auto path_str = tmp_base_dir.string();  // Copy the underlying path to get access to the raw char *.
+
+        // LCOV_EXCL_START
+        // Creating a test for this requires to create an artificial environment in which this program has no
+        // permission to write to the /tmp filesystem. This is outside of the scope of a unit test.
+        if (char * f = mkdtemp(path_str.data()); f == nullptr)  // mkdtemp replaces XXXXXXXX in a safe and unique way.
+        {
+            throw std::filesystem::filesystem_error("Could not create temporary directory with mkdtemp!",
+                                                    tmp_base_dir,
+                                                    std::make_error_code(std::errc::bad_file_descriptor));
+        }
+        // LCOV_EXCL_START
+        directory_path = std::make_unique<sandboxed_path>(path_str);
+    }
+
+    /*!\brief Destructs the temporary directory path.
+     *
+     * \details
+     *
+     * Removes the temporary directory and all its subdirectories and files contained.
+     */
+    ~tmp_directory()
+    {
+        if (directory_path)
+        {
+            if (!exists(*directory_path))
+            {
+                std::cerr << "temporary directory "
+                          << *directory_path
+                          << " was deleted externally. This is discouraged program behaviour\n";
+                return;
+            }
+            if (!empty())
+            {
+                std::cerr << "temporary directory " << *directory_path << " has some files that should be deleted\n";
+                for (auto & p: std::filesystem::recursive_directory_iterator(*directory_path))
+                {
+                    std::cerr << "- " << p << "\n";
+                }
+            }
+            std::filesystem::remove_all(*directory_path);
+        }
+    }
+    //!\}
+
+
+    /*!\brief Returns a reference to the path object.
+     * \return seqan3::test::sandboxed_path containing the path of the file.
+     */
+    sandboxed_path const& get_path() const
+    {
+        assert(directory_path);
+
+        return *directory_path;
+    }
+
+    /*!\brief Check if the temporary directory is empty.
+     * \return True if the directory is empty, false otherwise.
+     */
+    bool empty() const
+    {
+        assert(directory_path);
+
+        return exists(*directory_path) && is_empty(*directory_path);
+    }
+
+    /*!\brief Removes all files from the temporary directory
+     */
+    void clean()
+    {
+        assert(directory_path);
+
+        for (auto & p: std::filesystem::directory_iterator(*directory_path))
+        {
+            std::filesystem::remove_all(p);
+        }
+    }
+
+private:
+    std::unique_ptr<sandboxed_path> directory_path;
```

Comment:
I haven't completely understood why this needs to be a `unique_ptr` :)

Reply:
Yes, good point.

A `tmp_directory` can have two states:
1. It is managing a sandboxed_path
2. It was moved from and it is not managing anything anymore.

To track if I am in the 1. or 2. state I need some helper variables.
I can think of 3 different solutions:
1. a `bool` tracking if it is still managing or not.
2. `std::optional<sandboxed_path>`  instead of having a boolean we wrap the `directory_path` into a optional which has a much nicer semantic than using a `bool`
3. `std::unique_ptr<sandboxed_path>` same as optional just with unique_ptr

I decided to go with solution 3...no particular reason. I actually think option 2 is a little more cleaner. Should I switch over?

Reply:
Yes, please, because as you can see, I couldn't see this semantic behaviour.